Add nested structure descriptions to params & return value.

danielgtaylor opened the pull request:
This change modifies the documentation to include structure examples for
both nested input parameters and nested service response values. Responses
are included for both clients and any resource action which returns a low-
level response. This is part of the work for #54.

This does **not** yet include documentation for each of the nested values
except for a basic Python type. I'm hoping to get some feedback on
this before including the full member documentation.

Additionally, this includes some basic tests for the new functionality
in the `docs.py` module. It also includes a minor naming change for
PEP8 (more of these are coming soon).

Example output:

![screen shot 2015-03-26 at 10 57 55 am](https://cloud.githubusercontent.com/assets/106826/6853526/ef09f87e-d3a8-11e4-9410-77a184feb5ab.png)

cc @jamesls @kyleknap

kyleknap commented:
I am glad that you added tests so we can start building those up. I feel that we need them now as we are document some of the more complex structures.

One thing that I would like seen done is if you add "bcdoc"-esque helper methods in. I feel like if I was making edits to the doc code I could easily forget a newline character or a closing bracket. If we had some helper methods to help clean it up that would be great. Even if it was like methods like ``start_structure()``, ``end_structure()``, ``start_list()``, ``end_list()`` that made starting and ending brackets that would be great. Looking at the code, other methods that come off of the top of my head would be a ``structure_element(key, value)`` that returns the `` 'key': 'value' `` pattern in the documentation. I will let you know if I think of any more possible helper methods.

Let me know what you think.

danielgtaylor commented:
@kyleknap please take another look. Note that the test failure is because this depends on boto/botocore#502.

I've cleaned up the code quite a bit, added some constants/string templates to lower the chances of human error and along with that Botocore change am now testing a more complete example of the output that should catch your concerns. Any other feedback?

kyleknap commented:
Thanks. Looks good. :ship:

jamesls commented:
Looks like we're waiting on the linked botocore PR to get merged so we can get a clean run through travis-ci.

Also, can you please update the .coveragerc file to stop omitting coverage reports from this?  Since we're moving forward with this approach we should have this as part of our coverage report.

jamesls commented on boto3/docs.py:
```diff
@@ -591,11 +592,122 @@ def document_operation(operation_model, service_name, operation_name=None,
+        # Convert the description from HTML to RST (to later be converted
+        # into HTML... don't ask). If the parameter is a nested structure
+        # then we also describe its members.
+        param_desc = html_to_rst(
```
We should test at least all the new parts we've added here.  I'm not seeing any tests for this part.

danielgtaylor replied:
This was just moved from down below. It was moved up and a comment added so that we can modify it before rendering. I can add a test and plan to cover the whole module eventually.

jamesls commented on boto3/docs.py:
```diff
@@ -591,11 +592,122 @@ def document_operation(operation_model, service_name, operation_name=None,
+        param_desc = html_to_rst(
+            value.documentation, indent=9, indent_first=True)
+        if param_type in ['list', 'dict']:
+            param_desc = ('\n         Structure description::\n\n' +
```
Why are we saying structure description?  Structure == dict/map.  Why is this checking for a list?

danielgtaylor replied:
Structure as in object structure. The structure of the parameter. I'm open to better words / phrasing. Suggestions?

jamesls commented:
I might be misunderstanding @kyleknap feedback, but it doesn't seem like his feedback was addressed.  It wasn't about using constants so we don't accidently typo what a `{` char looks like, it was having a separate layer that was responsible for what the specific RST syntax looked like so this doc module is all about the logic of *how* the documentation should be generated (without having to worry about the specific syntax).

In which case I'm strongly in favor of this approach.

The CLI mostly uses this approach and it's worked well.

At the very least we should adopt this style for the new code you've added.  We can try and go back and update the existing code.

jamesls commented on boto3/docs.py:
```diff
@@ -591,11 +592,122 @@ def document_operation(operation_model, service_name, operation_name=None,
+            None, shape.member, indent=indent + 4, eol=',\n')
+        docs += spaces + '    ' + ELLIPSIS
+        docs += spaces + LIST_END + eol
+    else:
```
You're not handling maps.

danielgtaylor replied:
Maps are handled in the `else` block and are not yet rendered out. I'm hoping to add that shortly but I wasn't completely sure of the format.

jamesls commented on boto3/docs.py:
```diff
@@ -591,11 +592,122 @@ def document_operation(operation_model, service_name, operation_name=None,
+            if i < len(shape.members) - 1:
+                member_eol = ',' + member_eol
+
+            docs += document_structure(
```
Doesn't look like you're handling recursive shapes.

danielgtaylor replied:
Nice catch. Everything seems to render fine but I'll update this to detect loops for the future.

kyleknap commented:
@jamesls 
The specific examples that I pointed out were addressed. But in general I agree with you, I feel the more abstraction there is the better when it comes to docs. For better abstraction, I could see additional helper methods like ``document_list()``, ``document_list_element()``, ``document_structure_element()`` added.

For me, I guess the level of abstraction should be based on if we are planning to reuse this code elsewhere and how long we plan on using it. I was under the impression this code is going to be relatively temporary.

If we are planning to reuse this code elsewhere, it would be useful to have a public documentation api that we could rely on such that if we refactor the internals of a public documentation method we can expect a similar output.

jamesls commented:
I think for me it comes down to the fact that while we initially thought this code was temporary, it's stuck around far longer than we thought and we keep adding to it.  At this point, I think we need to just do the due diligence and add the proper abstractions/tests code and pay off the technical debt.

Even if (when) we do end up refactoring all this, if we have separate and decoupled parts, we should be able to reuse and pull out all the RST generation code since it won't be dependent on any boto3 specific parts.  This would mean that any work that goes into cleaning this code up won't be lost.

We're also planning on hopefully using as much as we can for the botocore autogen'd reference docs, and potentially may even pull some of this code down into botocore.

danielgtaylor commented:
@jamesls @kyleknap please have another look. I've update the code to use helper functions, have added a dummy service and test a bunch of the output from the doc generator. The coverage is now near to where it was before (around 96%). Still lower but it's a good set of basic tests that we can grow over time.

I am not testing batch actions and a couple other minor things, so we'll want to add tests for that in the future. What do you guys think?

kyleknap commented on boto3/docs.py:
```diff
@@ -120,16 +120,20 @@ def html_to_rst(html, indent=0, indent_first=False):
     :param service_name: The service short-name, like 'ec2'
+    :type session: botocore.session.Session
+    :param session: Existing pre-setup session or ``None``.
     :rtype: string
```
Can you add a doc string for ``resource_filename``?

kyleknap commented:
Definitely looks better. It would still be nice to get coverage back to where it was, but It is a good start to build up our documentation methods and tests. :ship: We should look into adding tests for the missing lines and possibly more abstractions to the documentation methods in the near future.
